Parallel test (parallel='tests') does not work when priority is used …

krmahadevan opened the pull request:
…in Test

Fixes #1336

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

krmahadevan commented:
@juherr - This is what I was trying out locally which ensures that the test shared in #1336 now starts passing. But the overall TestNG build seems to have broken (I am pretty sure its due to my changes). So I was hoping to get some idea on whether I am progressing in the right direction for this.

cbeust commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -658,6 +659,11 @@ public long getTimeOut() {
 
+      @Override
+      public void setLatch(CountDownLatch latch) {
+
```
Maybe copy/paste the comment above saying this method is intentionally empty (which I really like).

krmahadevan replied:
Sure Cedric. I will do that. I don't think this PR is good enough to be merged. I was kind of looking for feedback around how to solve the problem that the issue was displaying. Since this involves the Graph part of TestNG I am trying to get more understanding. I thought I would first show my changes and check with you and @juherr on whether I am going in the right direction.

cbeust commented:
Got it, thanks @krmahadevan. Feel free to close it.

krmahadevan commented:
@cbeust / @juherr - Can you please help review this PR ? The failure is not related to my changeset and seems to be sporadic.

juherr commented on src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java:
```diff
@@ -61,20 +64,35 @@ public void run() {
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
```
Why only `TestMethodWorker`? We can suppose the issue is the same for the other workers.

krmahadevan replied:
I chose to stick to only `TestMethodWorker` because it's the smallest unit of work.

juherr replied:
Do you have test issues if you remove this check?

juherr replied:
Or just add `setLatch` on `TestMethodWorker` and cast.

krmahadevan replied:
I fixed this by removing the `instanceof` check.

juherr replied:
:+1:

juherr commented on src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java:
```diff
@@ -61,20 +64,35 @@ public void run() {
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```
If I understand the fix well, we are waiting for the end of all runnables.

What about using `awaitTermination()` instead of a `CountDownLatch`?

krmahadevan replied:
No we are not waiting for the end of all runnables. We are waiting for the completion of execution of all workers that are part of the freenodes. So for e.g., I have 3 methods for each priority (1, 2, 3), the latch would wait for all the 3 methods of priority 1 to run to completion, before it starts picking up methods of priority 2 and running them. Absence of this behavior was what was causing the tests to fail. 

awaitTermination, I believe is being down outside at `TestRunner` level, wherein the executor is being kicked off. This is more to do with kind of bringing in a discipline amongst tests that execute the `IWorker` variants for a given set of nodes. I hope I managed to explain it :)

juherr replied:
Sorry, but I'm not understanding the issue.

In case of 3x {1, 2, 3}, `Graph#getFreeNodes()` is supposed to only return {2} when all methods are `FINISHED`, right?
A `GraphThreadPoolExecutor#run()` only run free nodes and each run is on a new `GraphThreadPoolExecutor`.
We are waiting for the end of the execution after each run thanks to `awaitTermination` (which should be managed in the run method IMO).
What am I missing?

krmahadevan replied:
@juherr 

Please consider the test classes TestNG1, TestNG2, TestNG3 that are part of the the package [test.github1336](https://github.com/krmahadevan/testng/tree/8ad97dd15fe370cc272180455776cc19e73f8fbd/src/test/java/test/github1336).

For the first iteration : 
`org.testng.TestRunner#privateRun` > `org.testng.internal.thread.graph.GraphThreadPoolExecutor#run` > `org.testng.internal.thread.graph.GraphThreadPoolExecutor#runNodes`

The free-nodes would be as below : 

1. TestNG1.test1TestNG1()[pri:1, instance:test.github1336.TestNG1@23ab930d]
2. TestNG2.test1TestNG2()[pri:1, instance:test.github1336.TestNG2@4534b60d]
3. TestNG3.test1TestNG3()[pri:1, instance:test.github1336.TestNG3@3fa77460]

For the second iteration : 
`org.testng.internal.thread.graph.GraphThreadPoolExecutor#afterExecute` > `org.testng.internal.thread.graph.GraphThreadPoolExecutor#runNodes`

The free-nodes would be as below : 

Working with the freeNodes :
1. TestNG1.test2TestNG1()[pri:2, instance:test.github1336.TestNG1@23ab930d]
2. TestNG2.test2TestNG2()[pri:2, instance:test.github1336.TestNG2@4534b60d]
3. TestNG3.test2TestNG3()[pri:2, instance:test.github1336.TestNG3@3fa77460]
4. TestNG1.test3TestNG1()[pri:3, instance:test.github1336.TestNG1@23ab930d]
5. TestNG2.test3TestNG2()[pri:3, instance:test.github1336.TestNG2@4534b60d]
6. TestNG3.test3TestNG3()[pri:3, instance:test.github1336.TestNG3@3fa77460]


Without the latch, TestNG would immediately pick up the second set of free nodes without waiting for the first set of threads to run to completion.

Using an shutdown, should be done only after the all the threads have been submitted and not before that (its being done in `org.testng.internal.thread.graph.GraphThreadPoolExecutor#afterExecute` after the graph node has all nodes as completed ) and without invoking shutdown, there's no use of calling `awaitTermination`

juherr replied:
Ok, and hat about a less intrusive change like:

```java
  private void runNodes(List<T> freeNodes) {
    List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
    m_activeRunnables.addAll(runnables);
    List<Callable<Void>> callables = Lists.newArrayList(runnables.size());
    for (final IWorker<T> r : runnables) {
      setStatus(r, Status.RUNNING);
      callables.add(new Callable<Void>() {
        @Override
        public Void call() throws Exception {
          r.run();
          return null;
        }
      });
    }
    try {
      List<Future<Void>> futures = invokeAll(callables);
      for (Future<Void> future : futures) {
        try {
          future.get(10, TimeUnit.SECONDS);
        } catch (ExecutionException | TimeoutException e) {
          e.printStackTrace();
        }
      }
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }
```
?

BTW, I hope the second iteration is:

```
TestNG1.test2TestNG1()[pri:2, instance:test.github1336.TestNG1@23ab930d]
TestNG2.test2TestNG2()[pri:2, instance:test.github1336.TestNG2@4534b60d]
TestNG3.test2TestNG3()[pri:2, instance:test.github1336.TestNG3@3fa77460]
```

And the third one is:

```
TestNG1.test3TestNG1()[pri:3, instance:test.github1336.TestNG1@23ab930d]
TestNG2.test3TestNG2()[pri:3, instance:test.github1336.TestNG2@4534b60d]
TestNG3.test3TestNG3()[pri:3, instance:test.github1336.TestNG3@3fa77460]
```

(or we have an issue somewhere)

krmahadevan replied:
@juherr -Your approach would basically stall at every method for a max of 10 seconds no ?

w.r.t the iteration classification, yes, that was what I was hoping for as well, but its not like that. I still don't have enough understanding of `DynamicGraph` to be able to figure out where the problem is. Still trying to understand its logic.

juherr replied:
> Your approach would basically stall at every method for a max of 10 seconds no?

Yes and no. We can suppose they will be finished and won't take time. But we can imagine a more elegant solution.

> w.r.t the iteration classification, its not like that.

Ok, so I think it is the issue we have to fix first.
Are you able to add a new DynamicGraph test that will highlight the issue?

krmahadevan replied:
@juherr - The test that I added as part of this commit will essentially fail and can be considered as a test for `DynamicGraph`. On a side note, this is just a hunch. I don't think `DynamicGraph` basically distinguishes edges with the same weight on their priorities. I have fixed this with an additional commit.

> Yes and no. We can suppose they will be finished and won't take time. But we can imagine a more elegant solution.
I would like to understand the reason behind why you feel that the `Latch` based approach is not a befitting solution. Can you please help shed some light around that ?

juherr replied:
> I would like to understand the reason behind why you feel that the Latch based approach is not a befitting solution. Can you please help shed some light around that ?

I'm not feeling good to modify an interface with a method which has no sense in the domain when it should be an detail of implementation.

krmahadevan replied:
@juherr - Fair enough. I have now changed the logic and did the following 
* Extracted the latch adding behavior into its own interface that only `TestMethodWorker` implements. Now `GraphThreadPoolExecutor` runs an instance check and if it finds its an instance that can accept the latch, it goes on to set the latch to the worker. Please see if this would sound convincing.

krmahadevan commented on src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java:
```diff
@@ -26,18 +27,16 @@
   public GraphThreadPoolExecutor(DynamicGraph<T> graph, IThreadWorkerFactory<T> factory, int corePoolSize,
       int maximumPoolSize, long keepAliveTime, TimeUnit unit, BlockingQueue<Runnable> workQueue) {
-    super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue
-        /* , new TestNGThreadPoolFactory() */);
```
You want me to re-store the class as well (I deleted it off, because it wasn't being used) , or just the comment ?

juherr replied:
Yes, keep the class too. I think we should try to reuse it in another pr later.

juherr replied:
I dug and found how to fix the `TestNGThreadPoolFactory` issue: just change `TestNG` by something else. For example, "testng", "TestNg" or "testNG" are working. But "TestNG" anywhere in the thread name is failing.

I'd love to understand why "TestNG" is failing. 
@cbeust Any idea?

juherr replied:
Found!
https://github.com/cbeust/testng/blob/master/src/main/java/org/testng/internal/thread/ThreadUtil.java#L20-L27
Used by https://github.com/cbeust/testng/blob/master/src/main/java/org/testng/internal/MethodInvocationHelper.java#L241

juherr replied:
Fixed: #1350

cbeust replied:
I had totally forgotten I was doing things like that :-)

krmahadevan replied:
@juherr - I have rebased off of your changes that @cbeust  merged to master. So this is also taken care of.

juherr replied:
:+1:

krmahadevan commented:
@juherr - There was a problem with `DynamicGraph` wherein it was not additionally grouping same weighted edges based on their priority. This led to TestNG getting just two sets of free nodes (highest priority methods from where execution should start) and another set of free nodes (wherein they need to wait till the independent methods are executed). I have fixed this with an additional commit and also added tests for it.

juherr commented:
@krmahadevan Great! Nice catch! Could you make a new pull request with just the graph part?

krmahadevan commented:
@juherr - I have raised a separate PR for this https://github.com/cbeust/testng/pull/1361

krmahadevan commented:
@juherr - Can you please tell me what is pending on this PR ? I would like to wrap up all my open PRs before I begin picking up new issues.

ryudice commented:
Hi,
Is this fixed in version 6.14.3? I'm using that version and the issue still seems to happen

juherr commented:
@ryudice It is available since 6.12

devakash commented:
Any update ? Because Same issue is still happening in 6.14.3 :( 


`package com.omni.jtest;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

public class A {
	
	@BeforeClass
	public void init()
	{
		System.out.println("Before class a");
	}
	
	@Test(priority=1)
	public void a()
	{
		System.out.println("in method a");
	}
	
	@Test(priority=2)
	public void a2()
	{
		System.out.println("in method a2");
	}

}
`

package com.omni.jtest;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

public class B {
	
	@BeforeClass
	public void init()
	{
		System.out.println("Before class b");
	}
	
	@Test(priority=1)
	public void a()
	{
		System.out.println("in method b");
	}
	
	@Test(priority=2)
	public void a2()
	{
		System.out.println("in method b2");
	}

}


<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >

<suite name="Suites" parallel="tests" thread-count="2">

	
	<test name="Accounts2">
		<classes>
			<class name="com.omni.jtest.A" />
		
			<class name="com.omni.jtest.B" />
		</classes>

	</test>


</suite>

Actual output : 

Before class a
in method a
Before class b
in method b
in method a2
in method b2

Expected Output:

Before class a
in method a
in method a2
Before class b
in method b
in method b2

DhavalAtri commented:
I am also facing this issue with TestNG 6.14.3. 
Please reOpen this issue as it brakes the main functionality of our Testing framework.
Also Give Updates on the same.

juherr commented:
@DhavalAtri could you open a new issue instead?
